algebra/group/subgroup.lean

```diff
@@ -0,0 +1,147 @@
+import data.set.basic init.function data.equiv init.logic 
+import algebra.group
+
+open set
+
+universes u v
+variables {G : Type u} {H : Type v}
+
+class subgroup [group G] (S : set G) : Prop := 
+(mul_mem : ∀ {a b}, a ∈ S → b ∈ S → a * b ∈ S) 
+(one_mem : (1 : G) ∈ S)
+(inv_mem : ∀ {a}, a ∈ S → a⁻¹ ∈ S) 
+
+class normal_subgroup [group G] (S : set G) extends subgroup S : Prop :=
+(normal : ∀ n ∈ S, ∀ g : G, g * n * g⁻¹ ∈ S)
+             
+namespace subgroup
+variable [group G]
+
+-- Subgroup is a group
+instance (S : set G) [subgroup S] : group S := { 
+  mul := λ ⟨x, hx⟩ ⟨y, hy⟩, ⟨x * y, mul_mem hx hy⟩,
+  mul_assoc := λ ⟨x, hx⟩ ⟨y, hy⟩ ⟨z, hz⟩, subtype.eq $ mul_assoc x y z,
+  one := ⟨1, one_mem S⟩,
+  one_mul := λ ⟨x, hx⟩, subtype.eq $ one_mul x,
+  mul_one := λ ⟨x, hx⟩, subtype.eq $ mul_one x,
+  inv := λ ⟨x, hx⟩, ⟨x⁻¹, inv_mem hx⟩,
+  mul_left_inv := λ ⟨x, hx⟩, subtype.eq $ mul_left_inv x 
+}
+
+-- Normal subgroup properties
+lemma mem_norm_comm {a b : G} {S : set G} [normal_subgroup S] (hab : a * b ∈ S) : b * a ∈ S := 
+have h : a⁻¹ * (a * b) * a⁻¹⁻¹ ∈ S, from normal_subgroup.normal (a * b) hab a⁻¹,
+by simp at h; exact h
+
+-- Examples of subgroups
+@[simp]
+def trivial (G : Type u) [group G] : set G := {1}
```

Comment:
I don't think you want this definition to be `[simp]`, if you intend on using it at all. This will just make it unfold everywhere, making things like `trivial_normal` useless.

Reply:
Fixed.